[Serve][Doc] Direct Ingress (#29149)
ray-project/ray: doc/source/serve/direct-ingress.md

# Direct Ingress

> In the 2.1, Serve provides alpha version gRPC ingress. In this section, you will learn how to 

Reviewer: ```suggestion
In the 2.1, Serve provides an alpha version of gRPC ingress. In this section, you will learn how to 
```

> * use Serve internal gRPC schema to send traffic
> * bring your own gRPC schema into serve application

Reviewer: ```suggestion
* use Serve's internal gRPC schema to send traffic
* bring your own gRPC schema into your Serve application
```

> ## Use Serve Schema

Reviewer: ```suggestion
## Use Serve's Schema
```

> Internally, serve provides a simple gRPC schema.

Reviewer: ```suggestion
Internally, Serve provides a simple gRPC schema.
```

```
message PredictRequest {
  map<string, bytes> input = 2;
}

message PredictResponse {
  bytes prediction = 1;
}

service PredictAPIsService {
  rpc Predict(PredictRequest) returns (PredictResponse);
}
```

Code example:

> Server

Reviewer: ```suggestion
Server:
```

```{literalinclude} ../serve/doc_code/direct_ingress.py
:start-after: __begin_server__
:end-before: __end_server__
:language: python
```

> Client

Reviewer: ```suggestion
Client:
```

```{literalinclude} ../serve/doc_code/direct_ingress.py
:start-after: __begin_client__
:end-before: __end_client__
:language: python
```

:::{note}

> * `input` is a dictionary of `map<string, bytes> ` following the schema described

Reviewer: ```suggestion
* `input` is a dictionary of `map<string, bytes> ` following the schema described above.
```

Reviewer: In our discussion offline I remember we mentioned something like this is a standard interface where the strings are used for features.  Is it worth mentioning anything about that in the docs?

Reviewer: Cool, makes sense

> *  User input data needs to be serialized to bytes type and feed into the `input`.

Reviewer: ```suggestion
*  The user input data needs to be serialized to `bytes` type and fed into the `input`.
```

> * Response will be under bytes type, which means user code is responsible for to serialize the output into bytes. 

Reviewer: ```suggestion
* The response will be under `bytes` type, which means the user code is responsible for serializing the output into bytes. 
```

> * By default, the gRPC port is 9000.

Reviewer: Can we mention here how to change the default, or say that it can't be changed?

:::

### Client Schema code generation

> There are lots of ways to generate client schema code. Here is a simple step to generate the code.

Reviewer: Can you add a comment about when you need to do this? Do you always need to do it?  The user might be a little surprised because usually Ray features work out of the box without extra installation, so it might be worth adding a comment here

Author: When using RPC protocol, it is common that the user needs to have client code on the client side to construct the corresponding request which is outside Ray Serve scope.

Let's keep it as it is, if we receive lots of questions about it, and then let's update it. At least, it is a very common thing to me :)

Reviewer: Sounds good!

> * Insintall the gRPC code generation tools

Reviewer: ```suggestion
* Install the gRPC code generation tools
```

```
pip install grpcio-tools
```

* Generate gRPC code based on the schema
```
python -m grpc_tools.protoc --proto_path=src/ray/protobuf/ --python_out=. --grpc_python_out=. src/ray/protobuf/serve.proto
```

> After the two steps above, you should have `serve_pb2.py` and `serve_pb2_grpc.py` files generated.(The steps that show above work for all the schemas files generations.)

Reviewer: ```suggestion
After the two steps above, you should have `serve_pb2.py` and `serve_pb2_grpc.py` files generated.  (The steps shown above work for generation for any schema file, including "bring your own schema" described below.)
```

## Bring your own Schema

> If you have customized schema to use, serve also support it!

Reviewer: ```suggestion
If you have a customized schema to use, Serve also supports it!
```

> Assume you have the schema and generated the corresponding code,

Reviewer: ```suggestion
Assume you have the following custom schema and have generated the corresponding gRPC code:
```
Can we link to the code generation section here?

```
message PingRequest {
  bool no_reply = 1;
}
message PingReply {
}

message PingTimeoutRequest {}
message PingTimeoutReply {}

service TestService {
  rpc Ping(PingRequest) returns (PingReply);
  rpc PingTimeout(PingTimeoutRequest) returns (PingTimeoutReply);
}
```

Server:
```{literalinclude} ../serve/doc_code/direct_ingress_with_customized_schema.py
:start-after: __begin_server__
:end-before: __end_server__
:language: python
```

> Client

Reviewer: ```suggestion
Client:
```

> # Direct Ingress

Reviewer 2: ```suggestion
# gRPC Support
```

> In the 2.1, Serve provides an alpha version of gRPC ingress. In this section, you will learn how to

Reviewer 2: ```suggestion
In the 2.1, Serve provides an alpha version of [gRPC](https://grpc.io/) ingress. In this section, you will learn how to
```

> * use Serve's internal gRPC schema to send traffic

Reviewer 2: ```suggestion
* use Serve's builtin gRPC schema to receive client requests
```

* bring your own gRPC schema into your Serve application

> 

Reviewer 2: add a note saying this documentation assumes some basic understanding of gRPC

Reviewer 3: Or have a link to the gRPC documentation.

Author: added a grpc documentation above directly

## Use Serve's Schema

> Internally, Serve provides a simple gRPC schema.

Reviewer 2: ```suggestion
Serve provides a simple gRPC schema to machine learning inference workload. It is designed to be kept simple and you are encouraged to adapt it for your own need. 
```

Reviewer 3: It is designed to be kept simple, and you are encouraged to adapt it to your own need

```
message PredictRequest {
  map<string, bytes> input = 2;
}

message PredictResponse {
  bytes prediction = 1;
}

service PredictAPIsService {
  rpc Predict(PredictRequest) returns (PredictResponse);
}
```

> Code example:

Reviewer 2: ```suggestion
Take a look at the following code samples for using `DefaultgRPCDriver` in Ray Serve.
```

> Server:

Reviewer 2: ```suggestion
To implement the Serve, your class needs to inherit `ray.serve.drivers.DefaultgRPCDriver` and implement the `__call__` method.
```

```{literalinclude} ../serve/doc_code/direct_ingress.py
:start-after: __begin_server__
:end-before: __end_server__
:language: python
```

> Client:

Reviewer 2: You can use Serve's built-in gRPC client to send query to the model.

```{literalinclude} ../serve/doc_code/direct_ingress.py
:start-after: __begin_client__
:end-before: __end_client__
:language: python
```

:::{note}
* `input` is a dictionary of `map<string, bytes> ` following the schema described above.
*  The user input data needs to be serialized to `bytes` type and fed into the `input`.
* The response will be under `bytes` type, which means the user code is responsible for serializing the output into bytes.
* By default, the gRPC port is 9000. You can change it by passing port number when calling DefaultgRPCDriver bind function.
:::

> ### Client Schema code generation

Reviewer 2: ```suggestion
### Client schema code generation
```

> There are lots of ways to generate client schema code. Here is a simple step to generate the code.

Reviewer 2: ```suggestion
You can use the client either by importing it from the `ray` Python package. Alternatively, you can just copy [Serve's protobuf file](https://github.com/ray-project/ray/blob/e16f49b327bbc1c18e8fc5d0ac4fa8c2f1144412/src/ray/protobuf/serve.proto#L214-L225) to generate the gRPC client.
```

* Install the gRPC code generation tools
```
pip install grpcio-tools
```

* Generate gRPC code based on the schema
```
python -m grpc_tools.protoc --proto_path=src/ray/protobuf/ --python_out=. --grpc_python_out=. src/ray/protobuf/serve.proto
```

> After the two steps above, you should have `serve_pb2.py` and `serve_pb2_grpc.py` files generated.(The steps shown above work for generation for any schema file, including "bring your own schema" described below.)

Reviewer 2: ```suggestion
After the two steps above, you should have `serve_pb2.py` and `serve_pb2_grpc.py` files generated.
```

> ## Bring your own Schema

Reviewer 2: ```suggestion
## Bring your own schema
```

If you have a customized schema to use, Serve also supports it!

Assume you have the following customized schema and have generated the corresponding gRPC code:

```
message PingRequest {
  bool no_reply = 1;
}
message PingReply {
}

message PingTimeoutRequest {}
message PingTimeoutReply {}

service TestService {
  rpc Ping(PingRequest) returns (PingReply);
  rpc PingTimeout(PingTimeoutRequest) returns (PingTimeoutReply);
}
```

> Server:

Reviewer 2: Without Ray Serve, you can implement the business logic for gRPC server by creating a subclass of the generated `TestServiceServicer`. With Ray Serve, you can just needs two extra steps:
- Inherit `ray.serve.drivers.gRPCIngress` in your implementation class as well.
- Add the `@serve.deployment(is_driver_deployment=True)` decorator.
These two steps help Serve to manage your own gRPC server and schema.

```{literalinclude} ../serve/doc_code/direct_ingress_with_customized_schema.py
:start-after: __begin_server__
:end-before: __end_server__
:language: python
```

> Client:

Reviewer 2: You can now use your client unmodified.
